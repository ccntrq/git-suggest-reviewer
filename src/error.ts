export class GitCmdError extends Error {}
