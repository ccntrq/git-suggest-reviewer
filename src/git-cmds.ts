import {execSync} from 'child_process';

type SupportedGitCmd = 'diff' | 'blame';

export function gitBlame(
  baseRevision: string,
  file: string,
  startLine: number,
  lineCount: number
): string {
  return runGitCmd('blame', [
    baseRevision,
    file,
    `-L${startLine},+${lineCount}`,
  ]);
}

export function gitDiff(baseRevision: string): string {
  return runGitCmd('diff', [baseRevision]);
}

function runGitCmd(cmd: SupportedGitCmd, opts: Array<string>): string {
  const fullCmd = ['git', cmd, ...opts].join(' ');

  const result = execSync(fullCmd);

  return result.toString();
}
