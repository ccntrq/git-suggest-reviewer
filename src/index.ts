export * from './core';
export * from './errors';
export {ReviewerStats, renderTopReviewerTable} from './reviewer-stats';
