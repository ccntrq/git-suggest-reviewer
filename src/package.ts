const packageJson = require('../package.json');

export const version: string = packageJson.version;
export const bugsUrl: string = packageJson.bugs;
