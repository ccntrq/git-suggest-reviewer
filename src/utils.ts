export function lines(s: string): Array<string> {
  return s.split(/\n/);
}
