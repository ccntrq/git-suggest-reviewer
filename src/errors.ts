import {bugsUrl, version} from './package';

/**
 * An error of this class will be thrown if any of the git operations fail.
 */
export class GitCmdError extends Error {
  /**
   *
   * @param message Error message describing the git failure
   * @param originalError The original error that caused the failure
   */
  constructor(message: string, public readonly originalError: Error) {
    super(message);
  }
}

/**
 * Typeguard to check if an unknown error is an instance of  {@link GitCmdError | `GitCmdError`}.
 * @param error
 */
export function isGitCmdError(error: unknown): error is GitCmdError {
  return error instanceof Error && error.constructor.name === GitCmdError.name;
}

/**
 * Catch all error class to wrap unknown errors. An error of this class popping
 * up is considered a bug
 */
export class UnexpectedError extends Error {
  constructor(error: unknown) {
    super(
      'An unexpected error occured:\n' +
        (error instanceof Error && error.stack ? error.stack : `${error}`)
    );
  }

  /**
   * Use this method to generate a url for an issue containing relevant
   * information for this error.
   * @returns clickable url to create an issue for this error
   */
  newIssueUrl(): string {
    let url = `${bugsUrl}/new?title=`;

    url += encodeURIComponent('Unexpected Error occured');
    url += '&body=';
    url += encodeURIComponent(this.message + '\n\nVersion: ' + version);

    return url;
  }
}

/**
 * Typeguard to check if an unknown error is an instance of  {@link UnexpectedError | `UnexpectedError`}.
 * @param error
 */
export function isUnexpectedError(error: unknown): error is UnexpectedError {
  return (
    error instanceof Error && error.constructor.name === UnexpectedError.name
  );
}
