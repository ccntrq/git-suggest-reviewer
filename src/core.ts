import {GitCmdError, UnexpectedError} from './errors';
import {gitDiff, gitBlame} from './git-cmds';
import type {ReviewerStats, SuggestedReviewers} from './reviewer-stats';
import {lines} from './utils';

/**
 * Suggests possible reviewers based on the previous authors of changed lines
 * since `baseRevision`.
 *
 * @param baseRevision
 * @returns Array of {@link ReviewerStats}, sorted by the amount of changed
 * lines  per author in descending order.
 *
 * @throws {@link GitCmdError}
 * This exception is thrown in when any of the git operations fail
 *
 * @throws {@link UnexpectedError}
 * This exception is thrown on unknown/unhandled errors. This is considered a
 * bug.
 */
export function gitSuggestReviewer(baseRevision: string): SuggestedReviewers {
  try {
    return gitSuggestReviewerUnsafe(baseRevision);
  } catch (error: unknown) {
    if (
      error instanceof Error &&
      error?.constructor.name === GitCmdError.name
    ) {
      throw error;
    } else {
      throw new UnexpectedError(
        error instanceof Error && error.stack ? error.stack : `${error}`
      );
    }
  }
}

function gitSuggestReviewerUnsafe(baseRevision: string): SuggestedReviewers {
  const diff = gitDiff(baseRevision);
  const diffChanges = getDiffChanges(diff);
  const diffBlames = collectBlames(baseRevision, diffChanges);
  const diffBlameInfo = collectBlameInfo(diffBlames);
  const summary = summarizeBlameInfos([...diffBlameInfo.values()].flat());

  return summary;
}

type FilePath = string;

type DiffChanges = Map<FilePath, Array<Range>>;
type DiffBlames = Map<FilePath, Array<string>>;
type DiffBlameInfo = Map<FilePath, Array<BlameInfo>>;

interface BlameInfo {
  author: string;
  authorEmail: string;
  commitId: string;
  commitDate: Date;
}

interface Range {
  startLine: number;
  linesChanged: number;
}

function summarizeBlameInfos(blameInfos: Array<BlameInfo>): SuggestedReviewers {
  const map = new Map<string, ReviewerStats>();

  blameInfos.forEach(blameInfo => {
    const current = map.get(blameInfo.author);
    if (current) {
      current.changedLines += 1;
      if (compareDates(blameInfo.commitDate, current.lastCommitDate) > 0) {
        current.lastCommitDate = blameInfo.commitDate;
        current.lastCommitId = blameInfo.commitId;
      }
    } else {
      map.set(blameInfo.author, {
        author: blameInfo.author,
        authorEmail: blameInfo.authorEmail,
        lastCommitId: blameInfo.commitId,
        lastCommitDate: blameInfo.commitDate,
        changedLines: 1,
      });
    }
  });

  return [...map.values()].sort((a, b) => {
    return b.changedLines - a.changedLines;
  });
}

function collectBlameInfo(diffBlames: DiffBlames): DiffBlameInfo {
  const blameInfo: DiffBlameInfo = new Map();

  diffBlames.forEach((blames, file) => {
    const fileBlameInfo = blames.map(blame => {
      return parseBlame(blame);
    });

    blameInfo.set(file, fileBlameInfo.flat());
  });

  return blameInfo;
}

function collectBlames(baseRevision: string, changes: DiffChanges): DiffBlames {
  const blames: DiffBlames = new Map();

  changes.forEach((ranges, file) => {
    const fileBlames = ranges.map(range =>
      gitBlame(baseRevision, file, range.startLine, range.linesChanged)
    );

    blames.set(file, fileBlames);
  });

  return blames;
}

function getDiffChanges(diff: string): DiffChanges {
  const changes: DiffChanges = new Map();

  let currentfile: string | undefined = undefined;

  lines(diff).forEach(line => {
    const changedFileLine = line.match(/^--- a\/(.*)/);

    if (changedFileLine) {
      currentfile = changedFileLine[1];
      return; // next
    }

    const newFileLine = line.match(/^--- \/dev\/null/);
    if (newFileLine) {
      currentfile = undefined;
      return; // next
    }

    if (!currentfile) {
      // Currently in a file that didn't exist in baseRevision
      return; // next
    }

    const rangeLine = line.match(/^@@\s-(\d+)(?:,(\d+))?\s\+\d+,\d+\s@@/);
    if (rangeLine && rangeLine[1] && rangeLine[2]) {
      const range = {
        startLine: parseInt(rangeLine[1]),
        linesChanged: rangeLine[2] ? parseInt(rangeLine[2]) : 1,
      };
      const current = changes.get(currentfile);

      if (current) {
        // update existing array
        current.push(range);
      } else {
        changes.set(currentfile, [range]);
      }

      return; // next
    }
  });

  return changes;
}

function parseBlame(blame: string): Array<BlameInfo> {
  const chunks = blame.split(/\n\t.*\n/);

  return chunks.filter(p => !!p.length).map(parseBlameChunk);
}

// NOTE:
// Very brittle parsing. This will break if the output of git blame is not
function parseBlameChunk(chunk: string): BlameInfo {
  const getValue = (line: string): string => {
    const value = line.split(' ').slice(1).join(' ');
    if (!value) {
      throw new Error(`getValue failed: '${line}'`);
    }

    return value;
  };

  const removeAngleBrackets = (withAngleBrackets: string): string => {
    return withAngleBrackets.replace(/^<(.*)>$/, (_, m1) => m1);
  };

  const chunkLines = lines(chunk);
  const commitId = chunkLines[0]?.split(' ')[0];
  if (!(commitId && chunkLines[1] && chunkLines[2] && chunkLines[3])) {
    throw new Error(`Cannot parse blame: ${chunk}`);
  }

  return {
    author: getValue(chunkLines[1]),
    authorEmail: removeAngleBrackets(getValue(chunkLines[2])),
    commitId: commitId,
    commitDate: new Date(parseInt(getValue(chunkLines[3])) * 1000),
  };
}

function compareDates(a: Date, b: Date): number {
  return a.getTime() - b.getTime();
}
