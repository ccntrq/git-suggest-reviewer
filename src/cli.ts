import {gitSuggestReviewer, ReviewerStats} from './core';
import {isGitCmdError, isUnexpectedError, UnexpectedError} from './errors';
import {version} from './package';

export function cli(): void {
  const opts = parseArgs(process.argv);

  if (opts?.help) {
    console.log(usage());
    // eslint-disable-next-line no-process-exit
    process.exit(0);
  }

  if (opts?.version) {
    console.log(version);
    // eslint-disable-next-line no-process-exit
    process.exit(0);
  }

  if (!opts || !opts.baseRevision) {
    console.error(usage());
    // eslint-disable-next-line no-process-exit
    process.exit(1);
  }

  const baseRevision = opts.baseRevision;
  const topReviewers = handleAppErrors(() => gitSuggestReviewer(baseRevision));

  console.log(renderTopReviewerTable(topReviewers.slice(0, 9)));

  // eslint-disable-next-line no-process-exit
  process.exit(0);
}

export function usage() {
  const usage = `git-suggest-reviewers

Suggest candidates for a code review based on git history.

USAGE:
  git-suggest-reviewers [OPTIONS]
  git-suggest-reviewers base_revision

Options:
    -h --help       Show this help
    -v --version    Print program version
`;

  return usage;
}

/**
 * Creates a nicely formatted table from the given {@link ReviewerStats} for
 * printing on a terminal or other devices.
 *
 * @param topReviewers
 */
export function renderTopReviewerTable(
  topReviewers: Array<ReviewerStats>
): string {
  return renderTable(
    topReviewers.slice(0, 9),
    ['author', 'changedLines', 'lastCommitDate'],
    ['Author', 'Changed Lines', 'Last Commit Date']
  );
}

interface CliOptions {
  baseRevision?: string;
  help: boolean;
  version: boolean;
}

function defaultCliOptions(): CliOptions {
  return {
    help: false,
    version: false,
  };
}

function parseArgs(args: Array<string>): undefined | CliOptions {
  if (args.length < 3) {
    return undefined;
  }

  const options = defaultCliOptions();
  args.slice(2).forEach(arg => {
    switch (arg) {
      case '-h':
      case '--help':
        options.help = true;
        break;
      case '-v':
      case '--version':
        options.version = true;
        break;
      default:
        if (!arg.startsWith('-') && !options.baseRevision) {
          options.baseRevision = arg;
        } else {
          console.warn('Invalid argument: ' + arg);
        }
        break;
    }
  });

  return options;
}

function renderTable<T extends object>(
  values: Array<T>,
  accessors: Array<keyof T>,
  headings?: Array<string>
): string {
  if (headings && headings.length !== accessors.length) {
    throw new Error('Headings and accessors must have same length.');
  }
  const columns = accessors.map(a => values.map(value => String(value[a])));
  if (headings) {
    columns.forEach((c, i) => c.unshift(String(headings[i])));
  }
  const columnWidths: Array<number> = columns.map(column =>
    Math.max(...column.map(entry => entry.length))
  );

  return columns[0]
    .map((_, row) => {
      return columns
        .map((column, columnNumber) => {
          const val = column[row];
          return val + ' '.repeat(columnWidths[columnNumber] - val.length);
        })
        .join('  ');
    })
    .join('\n');
}

function handleAppErrors<T>(action: () => T): T {
  try {
    return action();
  } catch (error: unknown) {
    if (isGitCmdError(error)) {
      console.error(error.message);
      // eslint-disable-next-line no-process-exit
      process.exit(2);
    }

    console.error(
      'An unexpected error occured. Please create a bug report by clicking the\nfollowing link:\n' +
        (isUnexpectedError(error)
          ? error
          : new UnexpectedError(error)
        ).newIssueUrl()
    );
    // eslint-disable-next-line no-process-exit
    process.exit(3);
  }
}
