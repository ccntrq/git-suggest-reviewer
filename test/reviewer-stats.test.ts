import {describe, expect, test} from '@jest/globals';
import {renderTopReviewerTable} from '../src';
import type {SuggestedReviewers} from '../src/reviewer-stats';
import {lines} from '../src/utils';

describe('reviewer-stats', () => {
  test('renderTopReviewerTable works', () => {
    const reviewers: SuggestedReviewers = [
      {
        author: 'Peter Pan',
        authorEmail: 'peter@pan.de',
        changedLines: 87,
        lastCommitDate: new Date(Date.UTC(2022, 9, 23)),
        lastCommitId: 'ed7432a',
      },
      {
        author: 'Captain Hook',
        authorEmail: 'captain@hook.de',
        changedLines: 3,
        lastCommitDate: new Date(Date.UTC(2022, 9, 23)),
        lastCommitId: '1j4nn1d',
      },
    ];

    expect(renderTopReviewerTable(reviewers)).toBe(
      `Author        Changed Lines  Last Commit Date
Peter Pan     87             2022-10-23      
Captain Hook  3              2022-10-23      `
    );
  });
  test('renderTopReviewerTable limits output to 10 reviewers', () => {
    const reviewers: SuggestedReviewers = [
      {
        author: 'Peter Pan',
        authorEmail: 'peter@pan.de',
        changedLines: 87,
        lastCommitDate: new Date(Date.UTC(2022, 9, 23)),
        lastCommitId: 'ed7432a',
      },
      {
        author: 'Captain Hook',
        authorEmail: 'captain@hook.de',
        changedLines: 70,
        lastCommitDate: new Date(Date.UTC(2022, 9, 23)),
        lastCommitId: '1j4nn1d',
      },
      {
        author: 'Wendy Darling',
        authorEmail: 'wendy@darling.de',
        changedLines: 44,
        lastCommitDate: new Date(Date.UTC(2022, 9, 25)),
        lastCommitId: 'akakdn1',
      },
      {
        author: 'Tinker Bell',
        authorEmail: 'tinker@bell.de',
        changedLines: 37,
        lastCommitDate: new Date(Date.UTC(2019, 0, 21)),
        lastCommitId: 'akakdn1',
      },
      {
        author: 'Tiger Lily',
        authorEmail: 'tiger@lily.de',
        changedLines: 29,
        lastCommitDate: new Date(Date.UTC(2018, 1, 1)),
        lastCommitId: 'adjn11l',
      },
      {
        author: 'John Darling',
        authorEmail: 'john@darling.de',
        changedLines: 21,
        lastCommitDate: new Date(Date.UTC(2012, 1, 1)),
        lastCommitId: 'okf14km',
      },
      {
        author: 'Michael Darling',
        authorEmail: 'michael@darling.de',
        changedLines: 21,
        lastCommitDate: new Date(Date.UTC(2009, 1, 1)),
        lastCommitId: 'k1dnnm1',
      },
      {
        author: 'George Darling',
        authorEmail: 'george@darling.de',
        changedLines: 21,
        lastCommitDate: new Date(Date.UTC(2006, 1, 1)),
        lastCommitId: 'ak1nn1e',
      },
      {
        author: 'Mary Darling',
        authorEmail: 'mary@darling.de',
        changedLines: 20,
        lastCommitDate: new Date(Date.UTC(2006, 1, 1)),
        lastCommitId: 'k2n51l1',
      },
      {
        author: 'James Barrie',
        authorEmail: 'james@barrie.de',
        changedLines: 9,
        lastCommitDate: new Date(Date.UTC(1860, 4, 1937)),
        lastCommitId: '1dalnael',
      },
      {
        author: 'Alexander Pankoff',
        authorEmail: 'ccntrq@screenri.de',
        changedLines: 1,
        lastCommitDate: new Date(Date.UTC(2020, 2, 22)),
        lastCommitId: '1dalnael',
      },
    ];

    expect(reviewers.length).toBeGreaterThan(10);
    expect(lines(renderTopReviewerTable(reviewers)).length).toBe(11);
  });
});
