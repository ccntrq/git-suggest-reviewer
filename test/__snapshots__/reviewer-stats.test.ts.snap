// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reviewer-stats renderTopReviewerTable with showEmail works 1`] = `
"Author                          Changed Lines  Last Commit Date
Peter Pan <peter@pan.de>        87             2022-10-23      
Captain Hook <captain@hook.de>  3              2022-10-23      "
`;

exports[`reviewer-stats renderTopReviewerTable works 1`] = `
"Author        Changed Lines  Last Commit Date
Peter Pan     87             2022-10-23      
Captain Hook  3              2022-10-23      "
`;
